content/tutorials/git_authentication/index.Rmd

```diff
@@ -42,29 +42,49 @@ If you want to get some intuition about what you are doing, you will need to rea
 
 Enable two-factor authentication for your GitHub account, follow [these steps](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication#configuring-two-factor-authentication-using-a-totp-mobile-app).
 Another two-factor authentication app that is not mentioned on that website but easy to use is Google Authenticator.
+You probably already use a two-factor authentication app, just use that.
```

Comment:
(Just trying to imitate a user a limited knowledge of authentication:)
_I guess I may use two-factor authentication for my email, or was it for my computer login?  I don't know exactly, and I have absolutely no idea what authentication app I use there (I just followed the manual of IT).  Anyway, I'll just try the manual..._
- _scanning the QR code gives that the requested application cannot be started and that the 'barcode' can be wrong (I suppose we don't have this app on our smartphones, as we buy and install them ourselves...)_
- _with 'enter this text code', I get a code, but I have no idea where I should enter it? (I probably can't install a app for it myself, as I don't have admin rights on my laptop)_
- _I have to add a 6-digit code, but the recovery codes I got for the two-factor authentication of my my email, are 8-digit codes, so I suppose this won't work._

So yes, I suppose in the best case users know that they indeed use 2FA already, but I doubt if they know which authentication app they are using, and how to apply it to this situation without any further explanation.  Without searching for which app we are using at INBO or installing an authentication app myself, I don't get it enabled, so unless users are really motivated and the app can be installed without admin rights, they won't succeed either, I think.

So I would really advise to make it a step-by-step manual applied to the situation of INBO that simply can be followed by executing the described steps.  Don't expect users to choose, install and use an app themselves without any manual: at least recommend an option or explain how they can use the app that is already present on their computer.

Reply:
Mmmh, I tend to disagree with you on this one: (i) the choice of 2FA / TOTP app is up to the user, (ii) anyone who is savvy and brave enough to be using Git, should be able to figure this out. Our IT recommended Google Authenticator (which I mention as well in the text), but really, the choice is up to you. Many password managers include a TOTP generator on board.  I can add guidelines here for - say - Google Authenticator, but it would be foolish to follow them if you happen to use KeePassXC or 1password already.
But I can be convinced otherwise. I might also be biased because I use these things regularly. What do others think?

Reply:
I agree with the fact that tech savvy people should have the choice and use whatever they have installed already (we don't force people to use this manual!), but the majority of the INBO users haven't installed their own tools, they just use whatever they are taught or told to use.  (They are not computer scientists, it is our task to support them with that if we want them to learn using Github because we think it is useful!)

I thought the purpose of your manual was to make it easier for these less tech savvy people to also use Github to collaborate with their tech savvy teammates in for instance reports?  If people have to figure out part of the manual themselves anyway because it is not explained, what is the sense of making a manual specific for INBO anyway?  There are already a lot of manuals with general and sometimes rather vague explications on how to set up authentications (that don't always work out well), so if you can't give any additional value, why doing the effort of making another manual like that?  The purpose of a INBO manual should in my opinion at least be to have less people that need additional help (we should gain at least the time needed for making a manual in a lower need of support).  (Also, don't forget: by leaving the choice up to them, we should be able to give support on all the different apps.)

It is your choice of supporting https, and honestly, if I see the complexity, I get less and less convinced that this is the right choice: when moving to a new laptop, for ssh I needed nothing but the general email instructions sent by IT support when getting a new laptop to get it functioning again.  For https it seems a lot of struggling, even with your manual.  So try to convince me https is easier than ssh (by providing a good manual), up to now it feels the opposite for me, and I'm sure this struggle will certainly not convince collegues to start using Github.  ;-)

All other INBO manuals are very straightforward and easy to use because choices are made, so why not doing the same with this manual?  You can give alternative options in the full text below TL;DR, where you also give the alternative of ssh.

Reply:
> It is your choice of supporting https, and honestly, if I see the complexity, I get less and less convinced that this is the right choice: when moving to a new laptop, for ssh I needed nothing but the general email instructions sent by IT support when getting a new laptop to get it functioning again. For https it seems a lot of struggling, even with your manual. So try to convince me https is easier than ssh (by providing a good manual), up to now it feels the opposite for me, and I'm sure this struggle will certainly not convince collegues to start using Github. ;-)

OK, let's take this back one step. In my head the choice for https stems from (i) it's the GitHub default, (ii) it's what I've been thought when learning Git at INBO, (iii) as far as I know it is still [what's thought at INBO](https://inbo.github.io/git-course/course_rstudio.html#23_Clone_a_repo_to_work_locally). In that respect it seemed logical to me to follow the recommendations put forward by the `usethis` package and the book happy git with R. But your comment makes me unsure. I'm certainly not the person to decide this for everyone at INBO. So before I proceed, someone needs to give me a green light to go ahead with this or not. Also, I'd rather discuss these things in person.

I am OK with adding a step by step instruction that explains how to setup the 2FA and I suggest we explain it using the Google Authenticator app (?), but first I would like to hear confirmation or not from other people to go ahead with HTTPS.

Now that I looked up the INBO material about the [INBO Git Workshop](https://inbo.github.io/git-course/index.html), I also realize that some or all of the material in this tutorial might better be integrated there?

Reply:
> OK, let's take this back one step. In my head the choice for https stems from (i) it's the GitHub default, (ii) it's what I've been thought when learning Git at INBO, (iii) as far as I know it is still [what's thought at INBO](https://inbo.github.io/git-course/course_rstudio.html#23_Clone_a_repo_to_work_locally).

https was indeed what has been taught at INBO, but this was before 2FA was required by Github.  And I suppose this was not yet required a year ago, when I did a Github coaching, as I didn't come across difficulties at that time...

At this point I can't tell which is the easiest: apparently I never enabled 2FA in Github (and I never did the effort to figure it out), and I already installed ssh some years ago, so this was the way of the least effort in my case.  Probably there is not a lot of difference (it both needs a lot of configuration steps once) except for packages that use one or another method, and the easiest one for a user is generally the one that has the clearest step-by-step manual.  ;-)

(For now I have one package that relies on git2r (and an old version of git2rdata) to read and write to a git repo for starting r users, so I will have to adapt this if https will be used at INBO, but on the other hand I still need to write a clear manual on the installation of the authentication stuff...  So in the end I gain by just relying on your manual and adapting my package. ;-)  )

So I'm fine with either https/2FA or ssh, the most important in my opinion is to have a clear step-by-step manual that can just be followed.  If 2FA and the use of Google Authenticator is advised by people who know it and it has no conflicts with commonly used packages at INBO, than it is fine for me.  Except for me (and I am mainly arguing for a clear manual, not against the method itself!) there seems not a lot of objection against it in this PR, so go ahead with HTTPS (of ask some more Windows users for their opinion???).

> Now that I looked up the INBO material about the [INBO Git Workshop](https://inbo.github.io/git-course/index.html), I also realize that some or all of the material in this tutorial might better be integrated there?

Good idea!

Reply:
>https was indeed what has been taught at INBO, but this was before 2FA was required by Github. And I suppose this was not yet required a year ago, when I did a Github coaching, as I didn't come across difficulties at that time...

Indeed, before August 13, 2021 you could get away with a simple username + password.

> (For now I have one package that relies on git2r (and an old version of git2rdata) to read and write to a git repo for starting r users, so I will have to adapt this if https will be used at INBO, but on the other hand I still need to write a clear manual on the installation of the authentication stuff... So in the end I gain by just relying on your manual and adapting my package. ;-) )

For that case, I recommend moving from git2r to gert. The latter supports both HTTPS and SSH.

> more Windows users for their opinion???.

Happy to hear feedback from anyone.

Reply:
Tagging some Windows users who might have an opinion about this: @damianooldoni @pietervsd @RCinbo

Reply:
I remember having chosen SSH for ease of use with `git2r` a few years ago, but `git2r` should support HTTPS as well, see https://docs.ropensci.org/git2r/#clone-a-repository, showing output like `origin (https://github.com/ropensci/git2r)`.

Indeed, in my case the print method for a repo outputs an SSH URL:

```
Remote:   design_groundw_exploremodels @ origin (git@github.com:inbo/n2khab-mne-design.git)
```

Reply:
The argument for HTTPS seems well explained in https://docs.ropensci.org/gert/#should-i-use-https-or-ssh-remotes. As is the argument for using `gert`.

Reply:
Indeed - contrary to what I was thinking - git2r does support https. But, I would still go for gert. A quote from https://docs.ropensci.org/gert/#automatic-authentication:

> To authenticate with a remote in git2r, you often need to manually pass your credentials in every call to, e.g., git2r::clone(). This is always the case for an https remote and is often the case even for an ssh remote.

Reply:
> That is statement is rubbish in case of an ssh remote.

Yes, it is advised to use `ssh-agent` to provide the private key (and the passphrase if set), cf. https://happygitwithr.com/ssh-keys.html#add-key-to-ssh-agent and https://docs.github.com/en/authentication/connecting-to-github-with-ssh. The agent will remember the key only during the user session (OS level), so this is to be done again in the next login session before doing remote operations. If this is not done, you won't be offered some alternative:

```
$ git remote -v
origin	git@github.com:inbo/tutorials.git (fetch)
origin	git@github.com:inbo/tutorials.git (push)
$ ssh-add -D
All identities removed.
$ git fetch
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```